File: lsd-interceptors/src/test/java/io/lsdconsulting/interceptors/messaging/EventConsumerInterceptorTest.java
package io.lsdconsulting.interceptors.messaging;

import com.lsd.core.LsdContext;
import com.lsd.core.domain.MessageType;
import org.assertj.core.api.AssertionsForClassTypes;
import org.assertj.core.api.AssertionsForInterfaceTypes;
import org.junit.jupiter.api.Test;
import org.mockito.ArgumentCaptor;
import org.springframework.messaging.Message;
import org.springframework.messaging.MessageChannel;
import org.springframework.messaging.MessageHeaders;

import java.util.Map;

import static com.lsd.core.domain.ParticipantType.PARTICIPANT;
import static io.lsdconsulting.interceptors.common.Headers.HeaderKeys.SOURCE_NAME;
import static io.lsdconsulting.interceptors.common.Headers.HeaderKeys.TARGET_NAME;
import static java.nio.charset.StandardCharsets.UTF_8;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.*;

class EventConsumerInterceptorTest {
    private final ArgumentCaptor<com.lsd.core.domain.Message> messageCaptor = ArgumentCaptor.forClass(com.lsd.core.domain.Message.class);

    private final LsdContext lsdContext = spy(LsdContext.class);
    private final Message<byte[]> message = spy(Message.class);

    private final EventConsumerInterceptor underTest = new EventConsumerInterceptor(lsdContext);

    @Test
    void logInteraction() {
        given(message.getPayload()).willReturn("{\"key\":\"value\"}".getBytes(UTF_8));
        given(message.getHeaders()).willReturn(new MessageHeaders(Map.of(SOURCE_NAME.key(), "Source", TARGET_NAME.key(), "Target")));

        underTest.preSend(message, mock(MessageChannel.class));

        verify(lsdContext).capture(messageCaptor.capture());
        var capturedMessage = messageCaptor.getValue();

        AssertionsForClassTypes.assertThat(capturedMessage.getFrom()).isEqualTo(PARTICIPANT.called("Source"));
        AssertionsForClassTypes.assertThat(capturedMessage.getTo()).isEqualTo(PARTICIPANT.called("Target"));
        AssertionsForClassTypes.assertThat(capturedMessage.getLabel()).isEqualTo("Consume event");
        AssertionsForInterfaceTypes.assertThat(capturedMessage.getType()).isEqualTo(MessageType.ASYNCHRONOUS);
        AssertionsForClassTypes.assertThat(capturedMessage.getData().toString())
                .contains("timestamp: " + message.getHeaders().get("timestamp"))
                .contains("Source-Name: Source")
                .contains("Target-Name: Target")
                .contains("id: " + message.getHeaders().get("id"));
    }
}

File: lsd-interceptors/src/test/kotlin/io/lsdconsulting/interceptors/messaging/EventConsumerInterceptorTest.kt
package io.lsdconsulting.interceptors.messaging

import com.lsd.core.LsdContext
import com.lsd.core.domain.MessageType
import com.lsd.core.domain.ParticipantType
import io.lsdconsulting.interceptors.common.HeaderKeys
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import io.mockk.spyk
import org.apache.commons.lang3.RandomStringUtils.randomAlphanumeric
import org.assertj.core.api.AssertionsForInterfaceTypes.assertThat
import org.junit.jupiter.api.Test
import org.springframework.messaging.Message
import org.springframework.messaging.MessageChannel
import org.springframework.messaging.MessageHeaders
import java.nio.charset.StandardCharsets

internal class EventConsumerInterceptorTest {
    private val messageSlot = slot<com.lsd.core.domain.Message>()
    private val lsdContext = spyk<LsdContext>()
    private val message = spyk<Message<ByteArray>>()
    private val appName = randomAlphanumeric(10)

    private val underTest = EventConsumerInterceptor(lsdContext, appName)

    @Test
    fun logInteraction() {
        every { lsdContext.capture(capture(messageSlot))} returns Unit
        every { message.payload } returns "{\"key\":\"value\"}".toByteArray(StandardCharsets.UTF_8)
        every { message.headers } returns
            MessageHeaders(
                mapOf<String, Any>(
                    HeaderKeys.SOURCE_NAME.key() to "Source",
                    HeaderKeys.TARGET_NAME.key() to "Target"
                )
            )

        underTest.preSend(message, mockk<MessageChannel>())
        val (_, from, to, label, type, _, data) = messageSlot.captured
        assertThat(from).isEqualTo(ParticipantType.PARTICIPANT.called("Target"))
        assertThat(to).isEqualTo(ParticipantType.PARTICIPANT.called(appName))
        assertThat(label).isEqualTo("Consume event")
        assertThat(type).isEqualTo(MessageType.ASYNCHRONOUS)
        assertThat(data.toString())
            .contains("timestamp: " + message.headers["timestamp"])
            .contains("Source-Name: Source")
            .contains("Target-Name: Target")
            .contains("id: " + message.headers["id"])
    }
}
